### DISCUSSION TITLE
Device doesn't seem to connect to MQTT if TLS is enabled

### DISCUSSION THREAD

**Discussion by ghost** [arendst/Tasmota#14757] [Q&A] [answered]
I have a Shelly 2.5 with a weird behaviour. If I flash `tasmota.bin` standard file from the repository and configure MQTT without TLS, the device connects without issues and I can see the topics it creates.

I wanted to secure the communication by enabling TLS, so I've followed the procedure described here: [](https://tasmota.github.io/docs/Self-signed-Mosquitto/). I've compiled the code using PlatformIO-Core, updating both `user_config_override.h` and `$TASMOTAROOT/tasmota/tasmota_ca.ino`. The compilation ends correctly and I'm able to flash the resulting firmware to the Shelly. Once flashed I'm able to setup and configure it without issue and I see in the MQTT configuration screen that it has a new `MQTT TLS` check box. I swap my current Mosquitto configuration to TLS and configure Tasmota's MQTT screen according to it, as well as run the `TLSKey1` and `TLSKey2` commands inside the Console to set the certificate in the device.

Now, once all configuration its done and the device reboots, it simply just doesn't establish communication with the MQTT server. I don't seem to see any try or log in the Console and I don't know what to check next...

Also I've check logs from Mosquitto and doesn't seem to receive any petition from Shelly. I can connect without issues with a MQTT client having the SSL and certificate configured. I suppose its not the MQTT server itself, since I have the logs set to debug and I don't see any try from the Shelly, my guess is that the Shelly isn't even trying to communicate with the MQTT server, but don't really know why... 

Any idea on what could be wrong or how to get more info regarding the issue?

#### Build changes

- tasmota/tasmota_ca.ino: Added the output of brssl ta <ca_file> between `const br_x509_trust_anchor PROGMEM AmazonRootCA1_TA = {...};` and `// cumulative CA` lines.
- tasmota/user_config_overrride.h
```
// !!! Remember that your changes GOES AT THE BOTTOM OF THIS FILE right before the last #endif !!!
*/

#ifndef USE_MQTT_TLS
#define USE_MQTT_TLS
#define USE_MQTT_TLS_CA_CERT                   // Force full CA validation instead of fingerprints, slower, but simpler to use.  (+2.2k code, +1.9k mem during connection handshake)
#define USE_MQTT_AWS_IOT    // This will include LetsEncrypt CA, as well as our CA, in tasmota_ca.ino for verifying server certificates
#define USE_MQTT_TLS_FORCE_EC_CIPHER           // Force Elliptic Curve cipher (higher security) required by some servers (automatically enabled with USE_MQTT_AWS_IOT) (+11.4k code, +0.4k mem)
#endif

#endif  // _USER_CONFIG_OVERRIDE_H_
```
#### Build/Flash procedure
I've tried building with either PlatformIO-Core (virtualenv) and Gitpod, both compile without issues. For flashing I use `esptool.py`, running `erase_flash` first, then `write_flash --flash_size 1MB --flash_mode dout 0x00000 firmware.bin`. Flash procedure works correctly and I'm able to boot the new flashed Shelly.

#### Configuration
- First configuration by connecting to the SSID the Shelly creates and configuring the WiFi.
- Once rebooted with the new configuration, set template, configure `TlsKey1` and `TlsKey2` from Console and configure MQTT server.

### MQTT information
![image](https://user-images.githubusercontent.com/44871070/152765093-d90780cb-e13b-4afe-9374-f6bd827ffb5b.png)

### Console log
```
00:00:00.001 HDW: ESP8266EX
00:00:00.050 CFG: Loaded from flash at F9, Count 19
00:00:00.055 QPC: Count 1
00:00:00.056 CFG: CR 374/699, Busy 0
00:00:00.061 CFG: CR 374/699, Busy 0
00:00:00.063 ROT: Mode 1
00:00:00.165 I2C: ADE7953 found at 0x38
00:00:00.166 SRC: Restart
00:00:00.168 Project tasmota - Shelly Version 10.1.0(tasmota)-2_7_4_9(2022-02-06T22:09:26)
00:00:00.221 WIF: Checking connection...
00:00:00.222 WIF: Attempting connection...
00:00:00.559 WIF: Connecting to AP1 <SID> Channel 6 BSSId XX:XX:XX:XX:XX in mode 11n as shelly-1171...
00:00:01.753 WIF: Checking connection...
00:00:01.754 WIF: Connected
00:00:02.006 HTP: Web server active on shelly-1171 with IP address X.Y.Z.A
00:00:02.403 WIF: Sending Gratuitous ARP
00:00:02.405 NTP: Synch time...
00:00:02.534 RTC: UTC 2022-02-07T09:45:33, DST 2022-03-27T02:00:00, STD 2022-10-30T03:00:00
10:45:33.001 NTP: Synched
10:45:34.048 RSL: INFO1 = {"Info1":{"Module":"Shelly 2.5","Version":"10.1.0(tasmota)","FallbackTopic":"cmnd/shelly-1_fb/","GroupTopic":"cmnd/tasmotas/"}}
10:45:34.053 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"shelly-1171","IPAddress":"X.Y.Z.A"}}
10:45:34.063 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart"}}
10:45:34.070 RSL: RESULT = {"POWER1":"OFF"}
10:45:34.073 RSL: POWER1 = OFF
10:45:34.076 RSL: RESULT = {"POWER2":"OFF"}
10:45:34.080 RSL: POWER2 = OFF
10:45:34.410 ADE: 1792, 63859, [1773, -1, 0, 0], [1773, 0, 0, -1]
10:45:34.411 ADE: U 1792, C 63859, I 0 + 0 = 0, P 0 + 0
10:45:35.410 ADE: 1792, 63859, [1773, -1, 0, 0], [1773, 0, 0, -1]
10:45:35.411 ADE: U 1792, C 63859, I 0 + 0 = 0, P 0 + 0
10:45:36.410 ADE: 1792, 63859, [1773, -1, 0, 0], [1773, 0, 0, -1]
10:45:36.411 ADE: U 1792, C 63859, I 0 + 0 = 0, P 0 + 0
10:45:37.410 ADE: 1792, 63859, [1773, -1, 0, 0], [1773, 0, 0, -1]
10:45:37.411 ADE: U 1792, C 63859, I 0 + 0 = 0, P 0 + 0
10:45:37.495 QPC: Reset
10:45:38.410 ADE: 1792, 63859, [1773, -1, 0, 0], [1773, 0, 0, -1]
10:45:38.411 ADE: U 1792, C 63859, I 0 + 0 = 0, P 0 + 0
10:45:38.418 RSL: STATE = {"Time":"2022-02-07T10:45:38","Uptime":"0T00:00:09","UptimeSec":9,"Heap":23,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER1":"OFF","POWER2":"OFF","Wifi":{"AP":1,"SSId":"<SID>","BSSId":"XX:XX:XX:XX:XX","Channel":6,"Mode":"11n","RSSI":76,"Signal":-62,"LinkCount":1,"Downtime":"0T00:00:03"}}
10:45:38.446 RSL: SENSOR = {"Time":"2022-02-07T10:45:38","Switch1":"OFF","Switch2":"OFF","ANALOG":{"Temperature":30.5},"ENERGY":{"TotalStartTime":"2022-02-06T22:12:46","Total":0.000,"Yesterday":0.000,"Today":0.000,"Period":[ 0, 0],"Power":[ 0, 0],"ApparentPower":[ 0, 0],"ReactivePower":[ 0, 0],"Factor":[0.00,0.00],"Voltage": 0,"Current":[0.000,0.000]},"TempUnit":"C"}
10:45:38.478 NRG: Energy reset by invalid data
10:45:38.888 HTP: Main Menu
10:45:39.374 APP: Boot Count 9
10:45:39.409 ADE: 1792, 63859, [1773, -1, 0, 0], [1773, 0, 0, -1]
10:45:39.411 ADE: U 1792, C 63859, I 0 + 0 = 0, P 0 + 0
10:45:39.622 CFG: Saved to flash at F8, Count 20, Bytes 4096
10:45:40.410 ADE: 1778, 63859, [1773, -1, 0, 0], [1773, 0, 0, -1]
10:45:40.412 ADE: U 1778, C 63859, I 0 + 0 = 0, P 0 + 0
10:45:41.410 ADE: 1778, 63859, [1773, -1, 0, 0], [1773, 0, 0, -1]
10:45:41.411 ADE: U 1778, C 63859, I 0 + 0 = 0, P 0 + 0
10:45:42.410 ADE: 1778, 63859, [1773, -1, 0, 0], [1773, 0, 0, -1]
10:45:42.412 ADE: U 1778, C 63859, I 0 + 0 = 0, P 0 + 0
10:45:42.721 HTP: Console
```

**Comment by s-hadinger** [comment, marked as answer]
If the server is Mosquitto, I don't see why you need TlsKey1/2. This is only for private certs.

Also you don't need `#define USE_MQTT_AWS_IOT` unless you connect to AWS IoT which does not sound like what you want.

I'm surprised that your device doesn't try to connect, which seems to indicate that something is missing in the configuration. One thing you can do is take another ESP8266 device and flash it with tasmota-zbbridge, this build already contains TLS.

**Reply by ghost** [reply]
I think I removed some link, I followed this guide: https://tasmota.github.io/docs/Self-signed-Mosquitto/

The TLS keys are needed to validate the device agains Mosquitto, since Mosquitto's SSL configuration requires each client to have its own certificate, so the server validates the certificate instead of using a username/password combination.

As per `USE_MQTT_AWS_IOT` I think its needed to include `tasmota_ca.ino`, but I haven't messed much with the code to know if I'm 100% right on this one.

I have a Sonoff in which I can test, but its a ESP32, so totally different... 😞

**Reply by s-hadinger** [reply]
It's a bit extreme to have a certificate for each device, especially on ESP8266 that has no secure storage. Can't you try first with TLS and user/password authentication?

**Reply by ghost** [reply]
Curious, if I set the username and password with my current configuration it fails, but at least it tries to contact the MQTT server:
```
18:25:38.043 MQT: Attempting connection...
18:25:38.352 MQT: TLS connection error: 62
18:25:38.354 MQT: Connect failed to mqtt.X.Y:8883, rc -2. Retry in 10 sec
```
Now I don't know why I get that error, regarding the documentation it seems it doesn't like my self-signed certificate despite its set in `tasmota_ca.ino`.

If I remove username/password from MQTT configuration it falls back to the previous behavior, it doesn't try to contact the server. I've done a test with just the username and its the same, so I suppose Tasmota requires both username and password in order to try to connect to the server, else it doesn't connect.

**Reply by s-hadinger** [reply]
Yes, user/password are required. If you are using a self-signed certificate, using a domain name that you haven't shared any details about, then you may be better using fingerprint validation.

**Reply by ghost** [reply]
By adding the CA to `tasmota_ca.ino`, I suppose the device should be able to validate the self-signed certificate. Its the same as if you install the CA in your browser and open a self-signed certificate server in your local lan, the browser checks that the certificate has been signed by the same CA and marks it as valid.

Aside that, I've tried the other method (the fingerprint) with the following lines in `user_config_override.h`:
```
#ifndef USE_MQTT_TLS
#define USE_MQTT_TLS
#define USE_MQTT_TLS_FORCE_EC_CIPHER           // Force Elliptic Curve cipher (higher security) required by some servers (automatically enabled with USE_MQTT_AWS_IOT) (+11.4k code, +0.4k mem)
#define MQTT_FINGERPRINT1      "00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00"  // [MqttFingerprint1] (auto-learn)
#define MQTT_FINGERPRINT2      "DA 39 A3 EE 5E 6B 4B 0D 32 55 BF EF 95 60 18 90 AF D8 07 09"  // [MqttFingerprint2] (invalid)
#endif
```

But configuring the same server ends in the same TLS error:
```
09:30:25.021 MQT: Attempting connection...
09:30:25.326 MQT: TLS connection error: 62
09:30:25.328 MQT: Connect failed to mqtt.X.Y:8883, rc -2. Retry in 10 sec
```

I'm not sure if this guide (https://tasmota.github.io/docs/Self-signed-Mosquitto) is up to date or not, but if you have any other guide/example to configure TLS without all this hassle, I will be happy to check it out.

**Reply by s-hadinger** [reply]
Error 62 means that the Server Certificate is not trusted, i.e. the CA is wrong. I don't know what you changed in `tasmota_ca.ino` but it may be wrong.

I doubt you enabled fingerprint validation, if so error 62 should not be possible since there is no check of the CA.

Happy to help but you need to share more about your configuration. It's hard to try guessing what you did and did not.

**Reply by ghost** [reply]
> Error 62 means that the Server Certificate is not trusted, i.e. the CA is wrong. I don't know what you changed in `tasmota_ca.ino` but it may be wrong.
> 
> I doubt you enabled fingerprint validation, if so error 62 should not be possible since there is no check of the CA.

I've added the lines described in my previous post in the `user_config_override.h`, then build and flash the resulting firmware. From the documentation I don't see there is more configuration needed, but please let me know which configuration you need and I'll post it.

**Reply by s-hadinger** [reply]
`SetOption132 1` forces fingerprint validation.

**Reply by ghost** [reply]
Ok, it seems we're progressing, now I get a TLS error 1, which seems to indicate a bad fingerprint:
```12:08:18.590 CMD: SetOption132
12:08:18.594 RSL: RESULT = {"SetOption132":"ON"}
12:08:21.008 MQT: Attempting connection...
12:08:21.091 MQT: TLS connection error: 1
12:08:21.094 MQT: Connect failed to mqtt.X.Y:8883, rc -2. Retry in 40 sec```

Do I have to reset the config or something like that to force the validation of the first fingerprint?

**Reply by s-hadinger** [reply]
Yes as described here: https://tasmota.github.io/docs/TLS/#fingerprint-validation

**Reply by ghost** [reply]
Same issue I'm afraid.

#### Steps done

1. Update `user_config_override.h` and add the following lines:
  ```
  #ifndef USE_MQTT_TLS 
  #define USE_MQTT_TLS                             // Use TLS for MQTT connection (+34.5k code, +7.0k mem and +4.8k additional during connection handshake)
  #define MQTT_FINGERPRINT1     "00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00"
  #endif
  ```
2. Compile the build (no errors), erase device and flash new build.
3. Restart device without GPIO, configure WiFi, device restarts on new WLAN.
4. Configure Shelly 2.5 template.
5. Configure MQTT server as follows:
![image](https://user-images.githubusercontent.com/44871070/153028963-a3a11077-4a37-4bf9-ac56-f596213096d0.png)
6. Once the device restarts, Console gives error TLS 62. Enter Console and run `SetOption132 1`.
7. Restart the device and once restarted it gives again error TLS 1:
![image](https://user-images.githubusercontent.com/44871070/153029421-56fce3a2-622a-487f-9e79-4407bf560075.png)

The guide on https://tasmota.github.io/docs/TLS states that you add `lib_extra_dirs          = lib/lib_ssl` in your `platformio_tasmota_env.ini`, but I wasn't able to compile the build with that option as it gives an error:
```
<...>/Tasmota/tasmota/xdrv_05_irremote.ino:62:29: fatal error: IRremoteESP8266.h: No such file or directory

 #include <IRremoteESP8266.h>
                             ^
compilation terminated.
Retrieved `.pio/build/tasmota/liba21/ESP8266WiFi/WiFiClientSecureAxTLS.cpp.o' from cache
*** [.pio/build/tasmota/src/tasmota.ino.cpp.o] Error 1
```
It seems its not able to find the header, but I'm not sure if I need to setup anything more in order to work.

**Reply by s-hadinger** [reply]
The `#define MQTT_FINGERPRINT1     "00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00"` only sets the default value but won't change what's in your settings.

What is the result of `MqttFingerprint1` and `MqttFingerprint2` commands? You can force auto-learn with `MqttFingerprint1 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00`

**Comment by Jason2866** [comment]
@s-hadinger I think i remember flashing the zbbridge build to a non zbbridge device ends in crashing (since hardware is not compatible).

**Reply by s-hadinger** [reply]
Oh, then don't try unless you can easily serial-reflash. Thanks @Jason2866
